### PR TITLE
IMPRO-974: Modify lightning threshold and update tests

### PR DESCRIPTION
**Pull request by bayliffe**
The purpose of this PR is to change the lightning probability threshold used in the weather symbol tree to 0.3 from 0.5. Also required was modifying the unit and acceptance tests to ensure this change is noticeable.

In addition I have removed a lot of duplicate cube set-up code. I have marked all the "new code" with comments, the rest is just slight refactoring.

Testing:
 - [x] Ran tests and they passed OK

### CODE REVIEW

File: improver/wxcode/wxcode_decision_tree.py [hunk 1]
```diff
@@ -82,7 +82,7 @@ def wxcode_decision_tree():
             'succeed': 'lightning_cloud',
             'fail': 'heavy_precipitation',
             'diagnostic_missing_action': 'fail',
-            'probability_thresholds': [0.5],
+            'probability_thresholds': [0.3],
```
Comment (bayliffe) [inline improver/wxcode/wxcode_decision_tree.py]: The fundamental change.

File: improver_tests/wxcode/wxcode/test_WeatherSymbols.py [hunk 2]
```diff
@@ -47,186 +47,118 @@
...
+                'rainfall_rate_in_vicinity',
+                'm s-1',
+                spp__relative_to_threshold='above',
+                forecast_thresholds=np.array([8.33333333e-09,
+                                              2.77777778e-08,
+                                              2.77777778e-07])))
+
+        data_lightning = np.zeros((1, 1, 3, 3))
+        data_lightning[0, 0, 0, 0] = 0.25
+        data_lightning[0, 0, 0, 1] = 0.30
```
Comment (bayliffe) [inline improver_tests/wxcode/wxcode/test_WeatherSymbols.py]: Changing some data to be non-zero and pick out the value of interest.

File: improver_tests/wxcode/wxcode/test_WeatherSymbols.py [hunk 3]
```diff
@@ -695,18 +624,22 @@ def test_removes_bounds(self):
...
         self.wxmeaning = " ".join(WX_DICT.values())
-        self.no_lightning_wxcode = np.array(
-            [[[1, 3, 5], [6, 7, 8], [10, 11, 12]]])
         self.expected_wxcode = np.array(
+            [[[1, 29, 5], [6, 7, 8], [10, 11, 12]]])
+        self.expected_wxcode_night = np.array(
+            [[[0, 28, 5], [6, 7, 8], [9, 11, 12]]])
+        self.expected_wxcode_no_lightning = np.array(
+            [[[1, 3, 5], [6, 7, 8], [10, 11, 12]]])
+        self.expected_wxcode_alternate = np.array(
```
Comment (bayliffe) [inline improver_tests/wxcode/wxcode/test_WeatherSymbols.py]: Names have been modified to better reflect what these actually are.

File: improver_tests/wxcode/wxcode/test_WeatherSymbols.py [hunk 4]
```diff
@@ -790,17 +720,16 @@ def test_weather_data(self):
         cubes[4].data = data_cloud
         cubes[5].data = data_cld_low
         cubes[6].data = data_vis
+        cubes[7].data = data_lightning
```
Comment (bayliffe) [inline improver_tests/wxcode/wxcode/test_WeatherSymbols.py]: Zeroed the lightning for the alternative data test to keep the expected values unchanged.

File: improver_tests/wxcode/wxcode/test_WeatherSymbols.py [hunk 5]
```diff
@@ -47,186 +47,118 @@
...
-    cubes_no_lightning = iris.cube.CubeList([snowfall_rate, rainfall_rate,
-                                             snowfall_vicinity,
-                                             rainfall_vicinity,
-                                             cloud, cloud_low,
-                                             visibility])
-
-    return cubes, cubes_no_lightning
-
-
-def set_up_wxcubes_global():
```
Comment (bayliffe) [inline improver_tests/wxcode/wxcode/test_WeatherSymbols.py]: Cubes all set up in one place now and the necessary ones for any given test are simply extracted from the complete list.

File: improver_tests/wxcode/wxcode/test_WeatherSymbols.py [hunk 6]
```diff
@@ -47,186 +47,118 @@
...
+                forecast_thresholds=np.array([0.0])))
+
+        self.cubes = iris.cube.CubeList([
+            snowfall_rate, rainfall_rate, snowfall_vicinity, rainfall_vicinity,
+            cloud, cloud_low, visibility, lightning])
+        all = [cube.name() for cube in self.cubes]
+        self.uk_no_lightning = [name for name in all
+                                if 'lightning' not in name]
+        self.gbl = [name for name in self.uk_no_lightning
+                    if 'vicinity' not in name]
```
Comment (bayliffe) [inline improver_tests/wxcode/wxcode/test_WeatherSymbols.py]: Lists of cubes to extract for tests that do not require all the cubes.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- improver_tests/wxcode/wxcode/test_WeatherSymbols.py  4
         

Clones removed
==============
+ improver_tests/wxcode/wxcode/test_WeatherSymbols.py  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4957787&bid=16382877)

File: improver_tests/wxcode/wxcode/test_WeatherSymbols.py [hunk 7]
```diff
@@ -47,186 +47,118 @@
...
+                data_lightning,
+                'number_of_lightning_flashes_per_unit_area_in_vicinity',
+                'm-2',
+                spp__relative_to_threshold='above',
+                forecast_thresholds=np.array([0.0])))
+
+        self.cubes = iris.cube.CubeList([
+            snowfall_rate, rainfall_rate, snowfall_vicinity, rainfall_vicinity,
+            cloud, cloud_low, visibility, lightning])
+        all = [cube.name() for cube in self.cubes]
```
Comment (benfitzpatrick) [inline improver_tests/wxcode/wxcode/test_WeatherSymbols.py]: Codacy Issue found: [Redefining built-in 'all'](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4957787)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/wxcode/wxcode/test_WeatherSymbols.py  4
         

Clones removed
==============
+ improver_tests/wxcode/wxcode/test_WeatherSymbols.py  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4957787&bid=16382877)

File: improver_tests/wxcode/wxcode/test_WeatherSymbols.py [hunk 8]
```diff
@@ -718,8 +651,7 @@ def test_basic(self):
         self.assertArrayEqual(result.attributes['weather_code'], self.wxcode)
         self.assertEqual(result.attributes['weather_code_meaning'],
                          self.wxmeaning)
-        self.assertArrayEqual(result.data,
-                              self.no_lightning_wxcode)
+        self.assertArrayEqual(result.data, self.expected_wxcode)
```
Comment (Katie-Howard) [inline improver_tests/wxcode/wxcode/test_WeatherSymbols.py]: Should we be using np.testing now?

File: improver_tests/wxcode/wxcode/test_WeatherSymbols.py [hunk 9]
```diff
@@ -47,186 +47,118 @@
...
+                             0.0, 0.0, 1.0, 1.0, 0.0, 0.0,
+                             0.0, 1.0, 0.0]).reshape((2, 1, 3, 3))
+        visibility = (
+            set_up_probability_threshold_cube(
+                data_vis,
+                'visibility_in_air',
+                'm',
+                spp__relative_to_threshold='below',
+                forecast_thresholds=np.array([1000.0, 5000.0])))
+
```
Comment (Katie-Howard) [inline improver_tests/wxcode/wxcode/test_WeatherSymbols.py]: Is there a reason there isn't the visibility coord anymore? (line 223 in the old code)
visibility.coord(var_name="threshold").attributes['spp_relative_to_threshold'] = 'below'

Comment (bayliffe) [reply improver_tests/wxcode/wxcode/test_WeatherSymbols.py]: Copy paste error. I removed the duplication of setting up the visibility cube, but kept the version without this relative to threshold attribute on the threshold coordinate. Re-added as it should be there. Good spot.

File: improver_tests/wxcode/wxcode/test_WeatherSymbols.py (hunk 8 above)
Comment (bayliffe) [reply improver_tests/wxcode/wxcode/test_WeatherSymbols.py]: We've not moved away from IrisTest yet as far as I am aware (I will ask around). We are being more careful with the "approximately equal to" stuff from IrisTest due to possibly inappropriate defaults, but the tests in here are exact equality, so I think they are fine.

Comment (Katie-Howard) [reply improver_tests/wxcode/wxcode/test_WeatherSymbols.py]: This is why I asked. I know it is being used, and I'm not sure what we should be using. It all works so I think this is fine for me 👍

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/wxcode/wxcode/test_WeatherSymbols.py  4
         

Clones removed
==============
+ improver_tests/wxcode/wxcode/test_WeatherSymbols.py  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4957787&bid=16382877)

**Comment by bayliffe** [comment]
@cgsandford I've modified the tests to use the set_up_test_cubes utility. I have set the time for the cubes to allow the tests to pass without changing the results; essentially forcing the time to the middle of the day in autumn. This particular time and date was required to get the whole default region used by set_up_test_cubes (that stretches up to a latitude of 80) into the day or night at the same time.

File: improver_tests/wxcode/wxcode/test_WeatherSymbols.py [hunk 10]
```diff
@@ -34,6 +34,7 @@
 
 import iris
 import numpy as np
+from datetime import datetime as dt
```
Comment (benfitzpatrick) [inline improver_tests/wxcode/wxcode/test_WeatherSymbols.py]: Codacy Issue found: [standard import "from datetime import datetime as dt" should be placed before "import iris"](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4957787)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- improver_tests/wxcode/wxcode/test_WeatherSymbols.py  4
         

Clones removed
==============
+ improver_tests/wxcode/wxcode/test_WeatherSymbols.py  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4957787&bid=16382877)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/wxcode/wxcode/test_WeatherSymbols.py  4
         

Clones removed
==============
+ improver_tests/wxcode/wxcode/test_WeatherSymbols.py  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4957787&bid=16382877)

**Pull request merged**